group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -367,6 +446,155 @@ public CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGro
             return result;
         }
 
+        public CompletableFuture<JoinGroupResponseData> sendGenericGroupJoin(
+            JoinGroupRequestData request
+        ) {
+            return sendGenericGroupJoin(request, false);
+        }
+
+        public CompletableFuture<JoinGroupResponseData> sendGenericGroupJoin(
+            JoinGroupRequestData request,
+            boolean requireKnownMemberId
+        ) {
+            return sendGenericGroupJoin(request, requireKnownMemberId, false, null);
+        }
+
+        public CompletableFuture<JoinGroupResponseData> sendGenericGroupJoin(
+            JoinGroupRequestData request,
+            boolean requireKnownMemberId,
+            boolean supportSkippingAssignment,
+            ExpectedGenericGroupResult expectedResult
+        ) {
+            // requireKnownMemberId is true: version >= 4
+            // supportSkippingAssignment is true: version >= 9
+            short joinGroupVersion = 3;
+
+            if (requireKnownMemberId) {
+                joinGroupVersion = 4;
+                if (supportSkippingAssignment) {
+                    joinGroupVersion = ApiKeys.JOIN_GROUP.latestVersion();
+                }
+            }
+
+            CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+
+            RequestContext context = new RequestContext(
+                new RequestHeader(
+                    ApiKeys.JOIN_GROUP,
+                    joinGroupVersion,
+                    "client",
+                    0
+                ),
+                "1",
+                InetAddress.getLoopbackAddress(),
+                KafkaPrincipal.ANONYMOUS,
+                ListenerName.forSecurityProtocol(SecurityProtocol.PLAINTEXT),
+                SecurityProtocol.PLAINTEXT,
+                ClientInformation.EMPTY,
+                false
+            );
+
+            CoordinatorResult<Void, Record> result = groupMetadataManager.genericGroupJoin(
+                context,
+                request,
+                responseFuture
+            );
+
+            if (expectedResult != null) {
+                GenericGroup group = groupMetadataManager.getOrMaybeCreateGenericGroup(
+                    request.groupId(),
+                    false
+                );
+
+                Record groupMetadataRecord;
+                if (expectedResult.isNewGroup) {
+                    groupMetadataRecord = newEmptyGroupMetadataRecord(group, MetadataVersion.latest());
+                } else {
+                    groupMetadataRecord = RecordHelpers.newGroupMetadataRecord(group, MetadataVersion.latest());
+                }
+
+                expectedResult.records = Collections.singletonList(groupMetadataRecord);
```

Comment:
i'll think a bit more on this as it will require a large change in this class.

one of the reasons i had it like this is that we mostly care about the responseFuture in the tests and wanted to hide the record/append future validations. The timer could also produce records which require setting things in advance.

i agree it is unclean, i'll address this in the next commit.